<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracción y Visualización de Datos de Acciones</h1>
<h2>Descripción</h2>


Extraer datos esenciales de un conjunto de datos y mostrarlos es una parte necesaria de la ciencia de datos; por lo tanto, las personas pueden tomar decisiones correctas basadas en los datos. En esta asignación, extraerás algunos datos de acciones y luego mostrarás estos datos en un gráfico.


<h2>Tabla de Contenidos</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Definir una Función que Haga un Gráfico</li>
        <li>Pregunta 1: Usar yfinance para Extraer Datos de Acciones</li>
        <li>Pregunta 2: Usar Webscraping para Extraer Datos de Ingresos de Tesla</li>
        <li>Pregunta 3: Usar yfinance para Extraer Datos de Acciones</li>
        <li>Pregunta 4: Usar Webscraping para Extraer Datos de Ingresos de GME</li>
        <li>Pregunta 5: Graficar Gráfico de Acciones de Tesla</li>
        <li>Pregunta 6: Graficar Gráfico de Acciones de GameStop</li>
    </ul>
<p>
    Tiempo Estimado Necesario: <strong>30 min</strong></p>
</div>

<hr>


***Nota***:- Si estás trabajando localmente usando anaconda, por favor descomenta el siguiente código y ejecútalo.
Usa la versión según tu versión de python.


In [ ]:
!pip install yfinance
!pip install bs4
!pip install nbformat
!pip install --upgrade plotly

In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [53]:
import plotly.io as pio
pio.renderers.default = "notebook"

En Python, puedes ignorar advertencias usando el módulo warnings. Puedes usar la función filterwarnings para filtrar o ignorar mensajes de advertencia específicos o categorías.


In [54]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Definir Función de Graficación


En esta sección, definimos la función `make_graph`. **No necesitas saber cómo funciona la función, solo debes preocuparte por las entradas. Toma un dataframe con datos de acciones (el dataframe debe contener columnas Date y Close), un dataframe con datos de ingresos (el dataframe debe contener columnas Date y Revenue), y el nombre de la acción.**


In [55]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()
    from IPython.display import display, HTML
    fig_html = fig.to_html()
    display(HTML(fig_html))

Usa la función make_graph que ya hemos definido. Necesitarás invocarla en las preguntas 5 y 6 para mostrar los gráficos y crear el dashboard. 
> **Nota: No necesitas redefinir la función para graficar gráficos en ningún otro lugar de este notebook; solo usa la función existente.**


## Pregunta 1: Usar yfinance para Extraer Datos de Acciones


Usando la función `Ticker` ingresa el símbolo ticker de la acción de la cual queremos extraer datos para crear un objeto ticker. La acción es Tesla y su símbolo ticker es `TSLA`.


In [37]:
tesla = yf.Ticker('TSLA')

Usando el objeto ticker y la función `history` extrae información de acciones y guárdala en un dataframe llamado `tesla_data`. Establece el parámetro `period` a `"max"` para que obtengamos información por la máxima cantidad de tiempo.


In [38]:
tesla_data = tesla.history(period = 'max')

**Reinicia el índice** usando la función `reset_index(inplace=True)` en el DataFrame tesla_data y muestra las primeras cinco filas del dataframe `tesla_data` usando la función `head`. Toma una captura de pantalla de los resultados y el código desde el comienzo de la Pregunta 1 hasta los resultados de abajo.


In [39]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

## Pregunta 2: Usar Webscraping para Extraer Datos de Ingresos de Tesla


Usa la librería `requests` para descargar la página web https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Guarda el texto de la respuesta como una variable llamada `html_data`.


In [40]:
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm'
html_data = requests.get(url).text

Parsea los datos html usando `beautiful_soup` usando parser, es decir, `html5lib` o `html.parser`.


In [41]:
soup_tesla = BeautifulSoup(html_data, 'html.parser')

Usando `BeautifulSoup` o la función `read_html` extrae la tabla con `Tesla Revenue` y almacénala en un dataframe llamado `tesla_revenue`. El dataframe debe tener columnas `Date` y `Revenue`.


<details><summary>Instrucciones paso a paso</summary>

```

Aquí están las instrucciones paso a paso:

1. Crear un DataFrame Vacío
2. Encontrar la Tabla Relevante
3. Verificar la Tabla de Ingresos Trimestrales de Tesla
4. Iterar a Través de las Filas en el Cuerpo de la Tabla
5. Extraer Datos de las Columnas
6. Agregar Datos al DataFrame

```
</details>


<details><summary>Haz clic aquí si necesitas ayuda localizando la tabla</summary>

```
    
Abajo está el código para aislar la tabla, ahora necesitarás hacer un bucle a través de las filas y columnas como en el laboratorio anterior
    
soup.find_all("tbody")[1]
    
Si quieres usar la función read_html, la tabla está ubicada en el índice 1

Nos estamos enfocando en los ingresos trimestrales en el laboratorio.
```

</details>


In [42]:
tesla_revenue = pd.DataFrame(columns=['Date', 'Revenue'])

for row in soup_tesla.find_all("tbody")[1].find_all('tr'):
    col = row.find_all("td")
    date = col[0].text
    rev = col[1].text
    
    tesla_revenue = pd.concat([tesla_revenue, pd.DataFrame({'Date':[date], 'Revenue':[rev]})],ignore_index=True)

tesla_revenue.head()

Date  Revenue
0  2022-09-30  $21,454
1  2022-06-30  $16,934
2  2022-03-31  $18,756
3  2021-12-31  $17,719
4  2021-09-30  $13,757

Ejecuta la siguiente línea para remover la coma y el signo de dólar de la columna `Revenue`. 


In [ ]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"",regex=True)

Ejecuta las siguientes líneas para remover cualquier cadena nula o vacía en la columna Revenue.


In [44]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Muestra las últimas 5 filas del dataframe `tesla_revenue` usando la función `tail`. Toma una captura de pantalla de los resultados.


In [45]:
tesla_revenue.tail()

Date Revenue
48  2010-09-30      31
49  2010-06-30      28
50  2010-03-31      21
52  2009-09-30      46
53  2009-06-30      27

## Pregunta 3: Usar yfinance para Extraer Datos de Acciones


Usando la función `Ticker` ingresa el símbolo ticker de la acción de la cual queremos extraer datos para crear un objeto ticker. La acción es GameStop y su símbolo ticker es `GME`.


In [26]:
gamestop = yf.Ticker('GME')

Usando el objeto ticker y la función `history` extrae información de acciones y guárdala en un dataframe llamado `gme_data`. Establece el parámetro `period` a `"max"` para que obtengamos información por la máxima cantidad de tiempo.


In [27]:
gme_data = gamestop.history(period = 'max')

**Reinicia el índice** usando la función `reset_index(inplace=True)` en el DataFrame gme_data y muestra las primeras cinco filas del dataframe `gme_data` usando la función `head`. Toma una captura de pantalla de los resultados y el código desde el comienzo de la Pregunta 3 hasta los resultados de abajo.


In [28]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  \
0 2002-02-13 00:00:00-05:00  1.620129  1.693350  1.603296  1.691667  76216000   
1 2002-02-14 00:00:00-05:00  1.712708  1.716074  1.670626  1.683251  11021600   
2 2002-02-15 00:00:00-05:00  1.683251  1.687459  1.658002  1.674834   8389600   
3 2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578047  1.607504   7410400   
4 2002-02-20 00:00:00-05:00  1.615920  1.662210  1.603296  1.662210   6892800   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

## Pregunta 4: Usar Webscraping para Extraer Datos de Ingresos de GME


Usa la librería `requests` para descargar la página web https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Guarda el texto de la respuesta como una variable llamada `html_data_2`.


In [29]:
url2= 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html'

html_data_2 = requests.get(url2).text

Parsea los datos html usando `beautiful_soup` usando parser, es decir, `html5lib` o `html.parser`.


In [31]:
soup_gme = BeautifulSoup(html_data_2, 'html.parser')

Usando `BeautifulSoup` o la función `read_html` extrae la tabla con `GameStop Revenue` y almacénala en un dataframe llamado `gme_revenue`. El dataframe debe tener columnas `Date` y `Revenue`. Asegúrate de que la coma y el signo de dólar sean removidos de la columna `Revenue`.


> **Nota: Usa el método similar a lo que hiciste en la pregunta 2.**  


<details><summary>Haz clic aquí si necesitas ayuda localizando la tabla</summary>

```
    
Abajo está el código para aislar la tabla, ahora necesitarás hacer un bucle a través de las filas y columnas como en el laboratorio anterior
    
soup.find_all("tbody")[1]
    
Si quieres usar la función read_html, la tabla está ubicada en el índice 1


```

</details>


In [ ]:
gme_revenue = pd.DataFrame(columns=['Date', 'Revenue'])

for row in soup_gme.find_all("tbody")[1].find_all('tr'):
    col = row.find_all("td")
    date = col[0].text
    rev = col[1].text
    
    gme_revenue = pd.concat([gme_revenue, pd.DataFrame({'Date':[date], 'Revenue':[rev]})],ignore_index=True)

gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace(',|\$',"",regex=True)
gme_revenue.dropna(inplace=True)
gme_revenue = gme_revenue[gme_revenue['Revenue'] != ""]


Muestra las últimas cinco filas del dataframe `gme_revenue` usando la función `tail`. Toma una captura de pantalla de los resultados.


In [34]:
gme_revenue.tail()

Date Revenue
57  2006-01-31    1667
58  2005-10-31     534
59  2005-07-31     416
60  2005-04-30     475
61  2005-01-31     709

## Pregunta 5: Graficar Gráfico de Acciones de Tesla


Usa la función `make_graph` para graficar los Datos de Acciones de Tesla, también proporciona un título para el gráfico. Nota que el gráfico solo mostrará datos hasta junio de 2021.


<details><summary>Pista</summary>

```

Solo necesitas invocar la función make_graph con el parámetro requerido para imprimir los gráficos. La estructura para llamar la función `make_graph` es `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


In [ ]:
make_graph(tesla_data, tesla_revenue, 'Tesla Inc. - Análisis de Rendimiento Financiero (2010-2021)')

## Pregunta 6: Graficar Gráfico de Acciones de GameStop


Usa la función `make_graph` para graficar los Datos de Acciones de GameStop, también proporciona un título para el gráfico. La estructura para llamar la función `make_graph` es `make_graph(gme_data, gme_revenue, 'GameStop')`. Nota que el gráfico solo mostrará datos hasta junio de 2021.


<details><summary>Pista</summary>

```

Solo necesitas invocar la función make_graph con el parámetro requerido para imprimir los gráficos. La estructura para llamar la función `make_graph` es `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


In [ ]:
make_graph(gme_data, gme_revenue, 'GameStop Corp. - Análisis de Rendimiento Financiero (2002-2021)')

<h2>Acerca de los Autores:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> tiene un PhD en Ingeniería Eléctrica, su investigación se enfocó en usar machine learning, signal processing, y computer vision para determinar cómo los videos impactan la cognición humana. Joseph ha estado trabajando para IBM desde que completó su PhD.

Azim Hirjani


## Registro de Cambios

| Fecha (YYYY-MM-DD) | Versión | Cambiado Por    | Descripción del Cambio        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Cambió la URL de GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Eliminó la parte Opcional |
| 2020-08-27        | 1.0     | Malika Singla | Agregó laboratorio a GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. Todos los derechos reservados. <h3/>

<p>
